jsf-ri/src/main/java/com/sun/faces/component/search/SearchExpressionHandlerImpl.java

```diff
@@ -0,0 +1,558 @@
+                    invokeKeywordResolvers(searchExpressionContext, previous, command, null, callback);
+                } else {
+
+                    if (facesContext.getApplication().getSearchKeywordResolver().isLeaf(searchExpressionContext, command)) {
+                        throw new FacesException("Expression cannot have keywords or ids at the right side: " + command);
+                    }
+
+                    final String finalRemainingExpression = remainingExpression;
+
+                    invokeKeywordResolvers(searchExpressionContext, previous, command, remainingExpression, new ContextCallback() {
+                        @Override
+                        public void invokeContextCallback(FacesContext facesContext, UIComponent target) {
+                            handler.invokeOnComponent(searchExpressionContext, target, finalRemainingExpression, callback);
+                        }
+                    });
+                }
+            } else {
+                UIComponent target = previous.findComponent(command);
+                if (target != null) {
+                    if (remainingExpression == null) {
+                        callback.invokeContextCallback(facesContext, target);
+                    } else {
+                        handler.invokeOnComponent(searchExpressionContext, target, remainingExpression, callback);
+                    }
+                }
+            } 
+        }
+        else {
+            UIComponent target = previous.findComponent(expression);
+            if (target != null) {
+                callback.invokeContextCallback(facesContext, target);
+            } else {
+                // fallback
+                // invokeOnComponent doesnt work with the leading ':'
+                char separatorChar = facesContext.getNamingContainerSeparatorChar();
+                if (expression.charAt(0) == separatorChar) {
+                    expression = expression.substring(1);
+                }
+
+                facesContext.getViewRoot().invokeOnComponent(facesContext, expression, callback);
+            }
+        }
+    }
+
+    protected void invokeKeywordResolvers(SearchExpressionContext searchExpressionContext, UIComponent previous,
+                             String command, String remainingExpression, ContextCallback callback)
+    {
+        // take the command and resolve it using the chain of responsibility pattern.
+        SearchKeywordContext searchContext = new SearchKeywordContext(searchExpressionContext, callback);
+        searchContext.setRemainingExpression(remainingExpression);
+
+        searchExpressionContext.getFacesContext().getApplication()
+                .getSearchKeywordResolver().resolve(searchContext, previous, command);
+    }
+    
+    @Override
+    public String[] splitExpressions(FacesContext context, String expressions) {
+        // we can't use a split(":") or split(" ") as keyword parameters might contain spaces or commas        
```

Comment:
Ahh, i see.
Yep, that would be but we would also need recursion because we have 2 or even more seperator chars. Not sure whats better readable - also not sure about performance.